lib/spack/spack/modules.py

```diff
@@ -259,3 +259,126 @@ def _write(self, m_file):
                 m_file.write("prepend-path %s \"%s\"\n" % (var, directory))
 
         m_file.write("prepend-path CMAKE_PREFIX_PATH \"%s\"\n" % self.spec.prefix)
+
+
+import spack.compilers
+
+
+class LmodModule(EnvModule):
+    name = 'lmod'
+    path = join_path(spack.share_path, "lmod", "modulefiles")
+
+    def __init__(self, spec=None):
+        super(LmodModule, self).__init__(spec)
+        # Sets tha appropriate category to be used with the 'family' function
+        if self.spec.name in spack.compilers.supported_compilers():
+            self.family = 'compiler'
+        elif self.spec.package.provides('mpi'):
+            self.family = 'mpi'
+        else:
+            self.family = None
+
+    @property
+    def file_name(self):
+        modulefiles_name = LmodModule.path
+        if self._use_system_compiler():
+            # If the module is installed using the system compiler put the modulefile in 'Core'
+            hierarchy_name = 'Core'
+        elif not self._is_mpi_dependent():
+            # If the module is serial and built using a compiler other than the system one,
+            # put the modulefile in '<Compiler>/<Version>'
+            hierarchy_name = self._compiler_module_directory(self.spec.compiler.name, self.spec.compiler.version)
+        elif self._is_mpi_dependent() and not self._use_system_compiler():
+            # If the module is MPI parallel then put the modulefile in '<MPI>/<MPI-Version>/<Compiler/<Compiler-Version>'
+            cname = self.spec.compiler.name
+            cversion = self.spec.compiler.version
+            for key, val in self.spec.dependencies.items():
```

Comment:
@alalazo: You can actually simplify this whole loop to this:
```
if not 'mpi' in spec:
    raise RuntimeError('No MPI dependency found')

mpi_name = spec['mpi'].name
mpi_version = spec['mpi'].version
```
The above is actually more correct, as well, because `spec.dependencies` is just the _direct_ deps.  If you want _all_ dependencies you want to look at `Spec.traverse()`, which will walk the DAG for you.  `Spec.__contains__` does this automatically and tests each dep to see if it satisfies the argument (which is why `'mpi' in spec` works).

Maybe this is better illustrated in an example:
```
>>> from spack.spec import Spec
>>> from spack.graph import graph_ascii

>>> s = Spec('adept-utils')
>>> s.normalize()
>>> graph_ascii(s)
o  adept-utils
|\
o |  mpi
 /
o  boost
>>> 'mpi' in s
True
>>> 'mpich' in s
False
>>> 'openmpi' in s
False

>>> s.concretize()
>>> graph_ascii(s)
o  adept-utils
|\
o |  mpich
 /
o  boost
>>> 'mpi' in s
True
>>> 'mpich' in s
True
>>> 'openmpi' in s
False
```